com.microsoft.java.test.plugin/src/main/java/com/microsoft/java/test/plugin/internal/JUnitTestSearcher.java

```diff
@@ -0,0 +1,175 @@
+import org.eclipse.jdt.core.ICompilationUnit;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IJavaProject;
+import org.eclipse.jdt.core.IMember;
+import org.eclipse.jdt.core.IMethod;
+import org.eclipse.jdt.core.IPackageFragmentRoot;
+import org.eclipse.jdt.core.IRegion;
+import org.eclipse.jdt.core.ISourceRange;
+import org.eclipse.jdt.core.ISourceReference;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeHierarchy;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.search.IJavaSearchConstants;
+import org.eclipse.jdt.core.search.IJavaSearchScope;
+import org.eclipse.jdt.core.search.SearchEngine;
+import org.eclipse.jdt.core.search.SearchMatch;
+import org.eclipse.jdt.core.search.SearchParticipant;
+import org.eclipse.jdt.core.search.SearchPattern;
+import org.eclipse.jdt.core.search.SearchRequestor;
+import org.eclipse.jdt.ls.core.internal.JDTUtils;
+import org.eclipse.jdt.ls.core.internal.ResourceUtils;
+import org.eclipse.lsp4j.Range;
+
+import com.microsoft.java.test.plugin.internal.testsuit.TestLevel;
+import com.microsoft.java.test.plugin.internal.testsuit.TestSuite;
+
+public class JUnitTestSearcher {
+	private final String JUNIT_TEST_ANNOTATION = "org.junit.Test";
+	private final String JUNIT_RUN_WITH_ANNOTATION = "org.junit.runner.RunWith";
+	
+	public List<TestSuite> searchAllTests(IProgressMonitor monitor) {
+		IRegion region = getRegion();
+		SearchPattern runWithPattern = SearchPattern.createPattern(
+				JUNIT_RUN_WITH_ANNOTATION,
+				IJavaSearchConstants.ANNOTATION_TYPE,
+				IJavaSearchConstants.ANNOTATION_TYPE_REFERENCE,
+				SearchPattern.R_EXACT_MATCH | SearchPattern.R_CASE_SENSITIVE);
+		SearchPattern testPattern = SearchPattern.createPattern(
+				JUNIT_TEST_ANNOTATION,
+				IJavaSearchConstants.ANNOTATION_TYPE,
+				IJavaSearchConstants.ANNOTATION_TYPE_REFERENCE,
+				SearchPattern.R_EXACT_MATCH | SearchPattern.R_CASE_SENSITIVE);
+		SearchPattern pattern = SearchPattern.createOrPattern(runWithPattern, testPattern);
+		List<TestSuite> tests = new ArrayList<>();
+		HashSet<IType> testClasses = new HashSet<>();
+
+		SearchRequestor requestor = new SearchRequestor() {
+			@Override
+			public void acceptSearchMatch( SearchMatch match ) throws CoreException {
+				
+				Object element = match.getElement();
+				if (element instanceof IType || element instanceof IMethod) {
+			        IMember member = (IMember) element;
+			        IType type =
+			            member.getElementType() == IJavaElement.TYPE
+			                ? (IType) member
+			                : member.getDeclaringType();
+			        testClasses.add(type);
+				}
+			}
+		};
+
+		try {
+			ITypeHierarchy hierarchy = JavaCore.newTypeHierarchy(region, null, null);
+		    IType[] allClasses = hierarchy.getAllClasses();
+		    IJavaSearchScope scope =
+		            SearchEngine.createJavaSearchScope(allClasses, IJavaSearchScope.SOURCES);

```

Comment:
If we can constrain the search scope to the project source files, it doesn't need the region any more.